analyzer/protocol/openvpn/openvpn.spicy

```diff
@@ -0,0 +1,181 @@
+# Copyright (c) 2021 by the Zeek Project. See LICENSE for details.
+#
+# Uses similar logic to https://github.com/corelight/zeek-openvpn
+
+module OpenVPN;
+
+import spicy;
+
+%byte-order = spicy::ByteOrder::Big;
+
+type Share = unit {
+  var handshake: HandshakeState;
+
+  on %init {
+    self.handshake = HandshakeState::BEGIN;
+  }
+};
+
+type HandshakeState = enum {
+	BEGIN,
+	CLIENT_RESET,
+	SERVER_RESET,
+	CLIENT_TLS,
+	SERVER_TLS
+};
+
+type Opcode = enum {
+	P_CONTROL_HARD_RESET_CLIENT_V1	= 0x01,
+	P_CONTROL_HARD_RESET_SERVER_V1	= 0x02,
+	P_CONTROL_SOFT_RESET_V1		= 0x03,
+	P_CONTROL_V1			= 0x04,
+	P_ACK_V1			= 0x05,
+	P_DATA_V1			= 0x06,
+	P_CONTROL_HARD_RESET_CLIENT_V2	= 0x07,
+	P_CONTROL_HARD_RESET_SERVER_V2	= 0x08,
+	P_DATA_V2			= 0x09
+};
+
+public type OpenVPNRecords = unit {
+	%context=Share;
+	records : OpenVPNRecord(self.context())[];
+};
+
+public type OpenVPNRecord = unit(inout share: Share) {
+	message_type: bitfield(8) {
+		opcode: 3..7 &convert=Opcode($$);
+		key_id: 0..2;
+	};
+
+	switch ( self.message_type.opcode ) {
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V1 -> control_hard_reset_client_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V1 -> control_hard_reset_server_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_SOFT_RESET_V1 -> control_soft_reset_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_V1 -> control_v1: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_ACK_V1 -> ack_v1: AckMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_DATA_V1 -> data_v1: DataMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_SERVER_V2 -> control_hard_reset_server_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_CONTROL_HARD_RESET_CLIENT_V2 -> control_hard_reset_client_v2: ControlMessage(self.message_type.opcode, self.message_type.key_id, False, share);
+		Opcode::P_DATA_V2 -> data_v2: DataMessage(self.message_type.opcode, self.message_type.key_id, True, share);
+	};
+};
+
+public type OpenVPNRecordsHMAC = unit {
+	%context=Share;
+	records : OpenVPNRecordHMAC(self.context())[];
+};
+
+public type OpenVPNRecordHMAC = unit(inout share: Share) {
```

Comment:
It looks like the only difference between `OpenVPNRecord` and `OpenVPNRecordHMAC` is the different boolean values that they pass to the events, is that right? If so, you could use just a single type and pass the boolean value as an additional parameter that the upper-layer units would pass in as desired.

Reply:
Correct, nice.  I missed that.